src/multirust-cli/main.rs

```diff
@@ -52,818 +50,44 @@ fn main() {
-                             .map(|b| format!("{:02x}", b))
-                             .join("");
-
-    // Download latest hash
-    let mut latest_hash = {
-        let hash_url = try!(utils::parse_url(&(url.clone() + ".sha256")));
-        let hash_file = try!(cfg.temp_cfg.new_file());
-        try!(utils::download_file(hash_url, &hash_file, None, ntfy!(&cfg.notify_handler)));
-        try!(utils::read_file("hash", &hash_file))
-    };
-    latest_hash.truncate(64);
-
-    // If up-to-date
-    if latest_hash == current_hash {
-        info!("Already up to date!");
-        return Ok(());
-    }
-
-    // Get download path
-    let download_file = try!(cfg.temp_cfg.new_file_with_ext("multirust-", env::consts::EXE_SUFFIX));
-    let download_url = try!(utils::parse_url(&url));
-
-    // Download new version
-    let mut hasher = Hasher::new(Type::SHA256);
-    try!(utils::download_file(download_url,
-                              &download_file,
-                              Some(&mut hasher),
-                              ntfy!(&cfg.notify_handler)));
-    let download_hash = hasher.finish()
-                              .iter()
-                              .map(|b| format!("{:02x}", b))
-                              .join("");
-
-    // Check that hash is correct
-    if latest_hash != download_hash {
-        return Err(Error::Install(multirust_dist::Error::ChecksumFailed {
-            url: url,
-            expected: latest_hash,
-            calculated: download_hash,
-        }));
-    }
-
-    // Mark as executable
-    try!(utils::make_executable(&download_file));
-
-    #[cfg(windows)]
-    fn inner(path: &Path) -> Result<()> {
-        let mut cmd = Command::new("cmd");
-        let _ = cmd.arg("/C")
-                   .arg("start")
-                   .arg(path)
-                   .arg("self")
-                   .arg("install")
-                   .arg("-m")
-                   .spawn();
-        Ok(())
-    }
-    #[cfg(not(windows))]
-    fn inner(path: &Path) -> Result<()> {
-        Ok(try!(utils::cmd_status("update",
-                                  Command::new(path).arg("self").arg("install").arg("-m"))))
-    }
-
-    println!("Installing...");
-    try!(inner(&download_file));
-    process::exit(0);
-}
-
-fn get_toolchain<'a>(cfg: &'a Cfg, m: &ArgMatches, create_parent: bool) -> Result<Toolchain<'a>> {
-    cfg.get_toolchain(m.value_of("toolchain").unwrap(), create_parent)
-}
-
-fn remove_toolchain_args(cfg: &Cfg, m: &ArgMatches) -> Result<()> {
-    try!(get_toolchain(cfg, m, false)).remove()
-}
-
-fn default_(cfg: &Cfg, m: &ArgMatches) -> Result<()> {
-    let toolchain = try!(get_toolchain(cfg, m, true));
-    if !try!(common_install_args(&toolchain, m)) {
-        if !toolchain.is_custom() {
-            try!(toolchain.install_from_dist_if_not_installed());
+    // The namem of arg0 determines how the program is going to behave
+    let arg0 = env::args().next().map(|a| PathBuf::from(a));
```

Comment:
Unfortunately clap doesn't support invalid unicode last time I checked, so it wouldn't help.